mlflow/store/artifact/cloud_artifact_repo.py

```diff
@@ -0,0 +1,252 @@
+    """
+    Computes the number of chunks to use for a multipart upload of the specified file.
+    """
+    return math.ceil(os.path.getsize(local_file) / chunk_size)
+
+
+def _complete_futures(futures_dict, file):
+    """
+    Waits for the completion of all the futures in the given dictionary and returns
+    a tuple of two dictionaries. The first dictionary contains the results of the
+    futures (unordered) and the second contains the errors (unordered) that occurred
+    during the execution of the futures.
+    """
+    results = {}
+    errors = {}
+
+    with ArtifactProgressBar.chunks(
+        os.path.getsize(file),
+        f"Uploading {file}",
+        _MULTIPART_UPLOAD_CHUNK_SIZE,
+    ) as pbar:
+        for future in as_completed(futures_dict):
+            key = futures_dict[future]
+            try:
+                results[key] = future.result()
+                pbar.update()
+            except Exception as e:
+                errors[key] = repr(e)
+
+    return results, errors
+
+
+class CloudArtifactRepository(ArtifactRepository):
+    def __init__(self, artifact_uri):
+        super().__init__(artifact_uri)
+        # Use an isolated thread pool executor for chunk uploads/downloads to avoid a deadlock
+        # caused by waiting for a chunk-upload/download task within a file-upload/download task.
+        # See https://superfastpython.com/threadpoolexecutor-deadlock/#Deadlock_1_Submit_and_Wait_for_a_Task_Within_a_Task
+        # for more details
+        self.chunk_thread_pool = self._create_thread_pool()
+
+    # Write APIs
+
+    def log_artifacts(self, local_dir, artifact_path=None):
+        """
+        Parallelized implementation of `log_artifacts`.
+        """
+        StagedArtifactUpload = namedtuple(
+            "StagedArtifactUpload",
+            [
+                # Local filesystem path of the source file to upload
+                "src_file_path",
+                # Base artifact URI-relative path specifying the upload destination
+                "artifact_file_path",
+            ],
+        )
+
+        artifact_path = artifact_path or ""
+
+        staged_uploads = []
+        for dirpath, _, filenames in os.walk(local_dir):
+            artifact_subdir = artifact_path
+            if dirpath != local_dir:
+                rel_path = os.path.relpath(dirpath, local_dir)
+                rel_path = relative_path_to_artifact_path(rel_path)
+                artifact_subdir = posixpath.join(artifact_path, rel_path)
+            for name in filenames:
+                src_file_path = os.path.join(dirpath, name)
+                src_file_name = os.path.basename(src_file_path)
+                staged_uploads.append(
+                    StagedArtifactUpload(
+                        src_file_path=src_file_path,
+                        artifact_file_path=posixpath.join(artifact_subdir, src_file_name),
+                    )
+                )
```

Comment:
This function is an exact copy of the original `DatabricksArtifactRepository.log_arifacts` except these lines. These lines were originally:
```
                file_path = os.path.join(dirpath, name)
                dst_run_relative_artifact_path = self._get_run_relative_artifact_path_for_upload(
                    src_file_path=file_path,
                    dst_artifact_dir=artifact_subdir,
                )
                staged_uploads.append(
                    StagedArtifactUpload(
                        src_file_path=file_path,
                        dst_run_relative_artifact_path=dst_run_relative_artifact_path,
                    )
                )
```
In `DatabricksArtifactRepository`,  `_get_run_relative_artifact_path_for_upload` essentially POSIX style joins `self.run_relative_artifact_repo_root_path`, `artifact_subdir`, and the base name of `src_file_path`. However, other repositories don't have `self.run_relative_artifact_repo_root_path` and only need `artifact_subdir` joined with the source file base name, which I call the `artifact_file_path`. 

In `CloudArtifactRepository.log_artifacts`, I've refactored the original code such that the `artifact_file_path` is calculated instead of `dst_run_relative_artifact_path` and is provided to the subclass `_upload_to_cloud` functions. Then, in the `DatabricksArtifactRepository` subclass, we join `self.run_relative_artifact_repo_root_path` with `artifact_file_path` anytime it is used, which is only when it is being sent to the backend. Hence, you'll see this join logic in `_get_write_credential_infos` and the AWS `_multipart_upload` function.